src/macaron/slsa_analyzer/ci_service/github_actions/github_actions_ci.py

```diff
@@ -624,3 +462,120 @@ def has_kws_in_log(self, latest_run: dict, build_log: list) -> bool:
+        ------
+        CallGraphError
+            Error raised when an error occurs while traversing the callgraph.
+        """
+        for node in callgraph.bfs():
+            # We are just interested in nodes that have bash commands.
+            if isinstance(node, BashNode):
+                # We collect useful contextual information for the called BashNode.
+                caller_node = node.caller
+                # The GitHub Actions workflow that triggers the path in the callgraph.
+                workflow_node = None
+                # The GitHub Actions job that triggers the path in the callgraph.
+                job_node = None
+                # The step in GitHub Actions job that triggers the path in the callgraph.
+                step_node = node if node.node_type == BashScriptType.INLINE else None
+
+                # Walk up the grallgraph to find the relevant caller nodes.
+                while caller_node is not None:
+                    match caller_node:
+                        case GitHubWorkflowNode():
+                            workflow_node = caller_node
+                        case GitHubJobNode():
+                            job_node = caller_node
+                        case BashNode(node_type=BashScriptType.INLINE):
+                            step_node = caller_node
+
+                    caller_node = caller_node.caller
+
+                # Check if there was an issue in finding any of the caller nodes.
+                if workflow_node is None or job_node is None or step_node is None:
+                    raise CallGraphError("Unable to traverse the call graph to find build commands.")
+
+                # Find the bash commands that call the build tool.
+                for cmd in node.parsed_bash_obj.get("commands", []):
+                    if build_tool.is_build_command(cmd):
+                        lang_versions = lang_distributions = lang_url = None
+                        if lang_model := find_language_setup_action(job_node, build_tool.language):
+                            lang_versions = lang_model.lang_versions
+                            lang_distributions = lang_model.lang_distributions
+                            lang_url = lang_model.lang_url
+                        yield BuildToolCommand(
+                            caller_path=workflow_node.source_path,
+                            ci_path=workflow_node.source_path,
+                            command=cmd,
+                            job_name=job_node.name,
+                            step_node=step_node,
+                            language=build_tool.language,
+                            language_versions=lang_versions,
+                            language_distributions=lang_distributions,
+                            language_url=lang_url,
+                            reachable_secrets=get_reachable_secrets(step_node),
+                            events=get_ci_events(workflow_node),
+                        )
```

Comment:
Any instance of `BuildToolCommand` already contains a reference to `step_node`. Other fields within `BuildToolCommand` are extracted from the current BashNode, the `workflow_node` and `job_node`. I wonder if instead for BuildToolCommand we only store the reference to those nodes and move the logic for extracting these information (e.g. build tool, language versions) inside that `BuildToolCommand` instance :thinking: ?
Or is there a reason we need to make BuildToolCommand a TypedDict that I wasn't aware of.

Reply:
The `language_versions`, `language_distributions`, etc. need to be resolved as part of the CI analysis. Delegating this resolution to the build tool creates a circular dependency.